### PR TITLE
KAFKA-10697: Remove ProduceResponse.responses

### PR DESCRIPTION
**Pull request by tang7526**
issue : https://issues.apache.org/jira/browse/KAFKA-10697

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by tang7526** [comment]
@chia7712 Could you help review this PR?  Thanks.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java [hunk 1]
```diff
@@ -40,42 +41,39 @@
...
-                buffer, version);
-
-        assertEquals(1, v5FromBytes.responses().size());
-        assertTrue(v5FromBytes.responses().containsKey(tp0));
-        ProduceResponse.PartitionResponse partitionResponse = v5FromBytes.responses().get(tp0);
-        assertEquals(100, partitionResponse.logStartOffset);
-        assertEquals(10000, partitionResponse.baseOffset);
+        ProduceResponse v5FromBytes = (ProduceResponse) AbstractResponse.parseResponse(ApiKeys.PRODUCE, buffer, version);
+
+        assertEquals(1, v5FromBytes.data().responses().size());
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java:57]: Is this equal to previous check? The previous `assertEquals` checks the number of `topic partition`. By contrast, it check the number of `topic` now.

**Review by chia7712 — COMMENTED** [review]
@tang7526 thanks for your patch. one major comment is left.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java (hunk 1 above)
Comment (tang7526) [reply clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java:57]: I split it into two `assertEquals`：
`assertEquals(1, v5FromBytes.data().responses().size());`
`assertEquals(1, topicProduceResponse.partitionResponses().size()); `

I think these two `assertEquals` equal to previous `assertEquals`.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 2]
```diff
@@ -155,10 +155,16 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
...
   }
 
-  @nowarn("cat=deprecation")
   val requestKeyToError = (topicNames: Map[Uuid, String]) => Map[ApiKeys, Nothing => Errors](
     ApiKeys.METADATA -> ((resp: requests.MetadataResponse) => resp.errors.asScala.find(_._1 == topic).getOrElse(("test", Errors.NONE))._2),
-    ApiKeys.PRODUCE -> ((resp: requests.ProduceResponse) => resp.responses.asScala.find(_._1 == tp).get._2.error),
+    ApiKeys.PRODUCE -> ((resp: requests.ProduceResponse) => {
+      Errors.forCode(
+        resp.data
+          .responses.asScala.find(_.name == topic).get
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: How about using `.responses.find(topic)`?

File: core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala [hunk 3]
```diff
@@ -337,12 +336,13 @@ class DynamicConnectionQuotaTest extends BaseRequestTest {
...
-  @nowarn("cat=deprecation")
   private def verifyConnection(socket: Socket): Unit = {
     val produceResponse = sendAndReceive[ProduceResponse](produceRequest, socket)
-    assertEquals(1, produceResponse.responses.size)
-    val (_, partitionResponse) = produceResponse.responses.asScala.head
-    assertEquals(Errors.NONE, partitionResponse.error)
+    assertEquals(1, produceResponse.data.responses.size)
+    val topicProduceResponse = produceResponse.data.responses.asScala.head
+    assertEquals(1, topicProduceResponse.partitionResponses.size)    
+    val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/EdgeCaseRequestTest.scala [hunk 4]
```diff
@@ -154,11 +153,12 @@ class EdgeCaseRequestTest extends KafkaServerTestHarness {
...
     assertEquals(correlationId, responseHeader.correlationId, "The correlationId should match request")
-    assertEquals(1, produceResponse.responses.size, "One partition response should be returned")
-
-    val partitionResponse = produceResponse.responses.get(topicPartition)
-    assertNotNull(partitionResponse)
-    assertEquals(Errors.NONE, partitionResponse.error, "There should be no error")
+    assertEquals(1, produceResponse.data.responses.size, "One topic response should be returned")
+    val topicProduceResponse = produceResponse.data.responses.asScala.head
+    assertEquals(1, topicProduceResponse.partitionResponses.size, "One partition response should be returned")    
+    val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/EdgeCaseRequestTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 5]
```diff
@@ -1455,10 +1453,11 @@ class KafkaApisTest {
...
       val response = capturedResponse.getValue.asInstanceOf[ProduceResponse]
 
-      assertEquals(1, response.responses().size())
-      for (partitionResponse <- response.responses().asScala) {
-        assertEquals(Errors.INVALID_PRODUCER_EPOCH, partitionResponse._2.error)
-      }
+      assertEquals(1, response.data.responses.size)
+      val topicProduceResponse = response.data.responses.asScala.head
+      assertEquals(1, topicProduceResponse.partitionResponses.size)   
+      val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala [hunk 6]
```diff
@@ -61,14 +59,17 @@ class ProduceRequestTest extends BaseRequestTest {
             .setAcks((-1).toShort)
             .setTimeoutMs(3000)
             .setTransactionalId(null)).build())
-      assertEquals(1, produceResponse.responses.size)
-      val (tp, partitionResponse) = produceResponse.responses.asScala.head
+      assertEquals(1, produceResponse.data.responses.size)
+      val topicProduceResponse = produceResponse.data.responses.asScala.head
+      assertEquals(1, topicProduceResponse.partitionResponses.size)   
+      val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala [hunk 7]
```diff
@@ -110,21 +110,25 @@ class ProduceRequestTest extends BaseRequestTest {
       .setAcks((-1).toShort)
       .setTimeoutMs(3000)
       .setTransactionalId(null)).build())
-    val (tp, partitionResponse) = produceResponse.responses.asScala.head
+
+    assertEquals(1, produceResponse.data.responses.size)
+    val topicProduceResponse = produceResponse.data.responses.asScala.head
+    assertEquals(1, topicProduceResponse.partitionResponses.size)   
+    val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala [hunk 8]
```diff
@@ -152,8 +156,11 @@ class ProduceRequestTest extends BaseRequestTest {
       .setTransactionalId(null)).build()
 
     val produceResponse = sendProduceRequest(nonReplicaId, produceRequest)
-    assertEquals(1, produceResponse.responses.size)
-    assertEquals(Errors.NOT_LEADER_OR_FOLLOWER, produceResponse.responses.asScala.head._2.error)
+    assertEquals(1, produceResponse.data.responses.size)
+    val topicProduceResponse = produceResponse.data.responses.asScala.head
+    assertEquals(1, topicProduceResponse.partitionResponses.size)   
+    val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala [hunk 9]
```diff
@@ -185,17 +191,20 @@ class ProduceRequestTest extends BaseRequestTest {
       .setAcks((-1).toShort)
       .setTimeoutMs(3000)
       .setTransactionalId(null)).build())
-    assertEquals(1, produceResponse.responses.size)
-    val (tp, partitionResponse) = produceResponse.responses.asScala.head
+
+    assertEquals(1, produceResponse.data.responses.size)
+    val topicProduceResponse = produceResponse.data.responses.asScala.head
+    assertEquals(1, topicProduceResponse.partitionResponses.size)   
+    val partitionProduceResponse = topicProduceResponse.partitionResponses.asScala.head;
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: redundant ";"

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala [hunk 10]
```diff
@@ -43,12 +42,11 @@ class ProduceRequestTest extends BaseRequestTest {
 
   val metricsKeySet = KafkaYammerMetrics.defaultRegistry.allMetrics.keySet.asScala
 
-  @nowarn("cat=deprecation")
   @Test
   def testSimpleProduceRequest(): Unit = {
     val (partition, leader) = createTopicAndFindPartitionWithLeader("topic")
 
-    def sendAndCheck(memoryRecords: MemoryRecords, expectedOffset: Long): ProduceResponse.PartitionResponse = {
+    def sendAndCheck(memoryRecords: MemoryRecords, expectedOffset: Long): ProduceResponseData.PartitionProduceResponse = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Could you remove the returned type if it is not used.

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala [hunk 11]
```diff
@@ -110,21 +110,25 @@ class ProduceRequestTest extends BaseRequestTest {
...
-    assertEquals(0, partitionResponse.recordErrors.get(0).batchIndex)
-    assertEquals(1, partitionResponse.recordErrors.get(1).batchIndex)
-    assertEquals(2, partitionResponse.recordErrors.get(2).batchIndex)
-    for (recordError <- partitionResponse.recordErrors.asScala) {
-      assertNotNull(recordError.message)
+    assertEquals(3, partitionProduceResponse.recordErrors.size)
+    assertEquals(0, partitionProduceResponse.recordErrors.get(0).batchIndex)
+    assertEquals(1, partitionProduceResponse.recordErrors.get(1).batchIndex)
+    assertEquals(2, partitionProduceResponse.recordErrors.get(2).batchIndex)
+    for (recordError <- partitionProduceResponse.recordErrors.asScala) {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Could you rewrite it by `foreach`?
```scala
partitionProduceResponse.recordErrors.asScala.foreach(recordError => assertNotNull(recordError.batchIndexErrorMessage))
```

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java [hunk 12]
```diff
@@ -40,42 +41,39 @@
...
+        ProduceResponseData.TopicProduceResponse topicProduceResponse = v5FromBytes.data().responses().iterator().next();
+        assertEquals(1, topicProduceResponse.partitionResponses().size());  
+        ProduceResponseData.PartitionProduceResponse partitionProduceResponse = topicProduceResponse.partitionResponses().iterator().next();
+        TopicPartition tp = new TopicPartition(topicProduceResponse.name(), partitionProduceResponse.index());
+        assertEquals(tp0, tp);
+
+        assertEquals(100, partitionProduceResponse.logStartOffset());
+        assertEquals(10000, partitionProduceResponse.baseOffset());
         assertEquals(10, v5FromBytes.throttleTimeMs());
-        assertEquals(responseData, v5Response.responses());
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java:61]: This check includes `logAppendTime`, `error code`, `error message` and `recordErrors`. Could you add them back?

**Review by chia7712 — COMMENTED** [review]
@tang7526 thanks for this patch. It address the great cleanup! Please take a look at following comments :)

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 2 above)
Comment (tang7526) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Done.

File: core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala (hunk 3 above)
Comment (tang7526) [reply core/src/test/scala/integration/kafka/network/DynamicConnectionQuotaTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/EdgeCaseRequestTest.scala (hunk 4 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/EdgeCaseRequestTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 5 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala (hunk 6 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala (hunk 8 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala (hunk 9 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala (hunk 10 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Done.

File: core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala (hunk 11 above)
Comment (tang7526) [reply core/src/test/scala/unit/kafka/server/ProduceRequestTest.scala]: Done.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java (hunk 12 above)
Comment (tang7526) [reply clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java:61]: Done.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java [hunk 13]
```diff
@@ -40,42 +42,42 @@
...
+        assertEquals(1, topicProduceResponse.partitionResponses().size());  
+        ProduceResponseData.PartitionProduceResponse partitionProduceResponse = topicProduceResponse.partitionResponses().iterator().next();
+        TopicPartition tp = new TopicPartition(topicProduceResponse.name(), partitionProduceResponse.index());
+        assertEquals(tp0, tp);
+
+        assertEquals(100, partitionProduceResponse.logStartOffset());
+        assertEquals(10000, partitionProduceResponse.baseOffset());
+        assertEquals(RecordBatch.NO_TIMESTAMP, partitionProduceResponse.logAppendTimeMs());
+        assertEquals(Errors.NONE, Errors.forCode(partitionProduceResponse.errorCode()));
+        assertEquals(null, partitionProduceResponse.errorMessage());
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java]: Could you use `assertNull`?

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java [hunk 14]
```diff
@@ -40,42 +42,42 @@
...
-        responseData.put(new TopicPartition("test", 0), new ProduceResponse.PartitionResponse(Errors.NONE,
-                10000, RecordBatch.NO_TIMESTAMP, 100));
+        responseData.put(new TopicPartition("test", 0), new ProduceResponse.PartitionResponse(Errors.NONE, 10000, RecordBatch.NO_TIMESTAMP, 100));
         ProduceResponse v0Response = new ProduceResponse(responseData);
         ProduceResponse v1Response = new ProduceResponse(responseData, 10);
         ProduceResponse v2Response = new ProduceResponse(responseData, 10);
         assertEquals(0, v0Response.throttleTimeMs(), "Throttle time must be zero");
         assertEquals(10, v1Response.throttleTimeMs(), "Throttle time must be 10");
         assertEquals(10, v2Response.throttleTimeMs(), "Throttle time must be 10");
-        assertEquals(responseData, v0Response.responses(), "Response data does not match");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java:76]: Could you add similar checks?

**Review by chia7712 — COMMENTED** [review]
@tang7526 thanks for updating code. a couple of comments. PTAL

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java (hunk 13 above)
Comment (tang7526) [reply clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java]: Done.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java (hunk 14 above)
Comment (tang7526) [reply clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java:76]: Done.

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java [hunk 15]
```diff
@@ -78,6 +80,20 @@ public void produceResponseVersionTest() {
         assertEquals(0, v0Response.throttleTimeMs(), "Throttle time must be zero");
         assertEquals(10, v1Response.throttleTimeMs(), "Throttle time must be 10");
         assertEquals(10, v2Response.throttleTimeMs(), "Throttle time must be 10");
+
+        List<ProduceResponse> arrResponse = Arrays.asList(v0Response, v1Response, v2Response);
+        for(ProduceResponse produceResponse:arrResponse) {
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java]: code style: `produceResponse : arrResponse`

**Review by chia7712 — APPROVED** [review]
@tang7526 thanks for your patch. +1 after addressing latest comment :)

File: clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java (hunk 15 above)
Comment (tang7526) [reply clients/src/test/java/org/apache/kafka/common/requests/ProduceResponseTest.java]: Done.

**Pull request merged**